rpc: add path to gethdkey

Sjors opened the pull request:
Depends on #26728.

This PR expands the wallet RPC `gethdkey` introduced in #26728. It takes a BIP32 path as argument and returns the xpub, along with the master key fingerprint.

To test this PR:

* create a regular descriptor wallet
* call `getxpub m/86h/0h/0h`
* call `listdescriptors`, compare the xpub in the `tr()` descriptor with the previous step

## Bigger picture

This paves the way for using Bitcoin Core as one signer in a multisig setup. It simplifies the proposed flow in #22067.

The eventual flow would be:
1. Create a blank wallet with a seed (either with no descriptors, or where its single key descriptors are not `active`, so they don't get used when calling `getnewaddress`). 
2. Call `getxpub m/87h/0h/0h` (as suggested in [BIP 87](https://github.com/bitcoin/bips/blob/master/bip-0087.mediawiki))
3. (Manually, with Specter or with a simple Python utility - TBD): craft a multisig descriptor containing this xpub
4. Call `importdescriptors` which will allow the import if its own fingerprint is recognized (and after checking the xpub)
5. The usual flow with `getnewaddress`, `send` and `walletprocesspsbt` (and their GUI equivalents)

This PR makes step (2) possible.

Step (1) would require a followup, because blank wallets currently can't have a seed. Step 4 would also require a followup,  so that `importdescriptors` treats an `xpub` as if it was an `xpriv`, after checking that it can derive the xpub from the seed (that matches the fingerprint)

On the GUI side a wizard could perform the above steps and guide the user, where they either:
* export their own xpub and then import a descriptor from somewhere else (performing some sanity checks); or
* import xpub(s) from other wallets, compose a multisig and get the descriptors for export; or
* get xpub(s) from connected hardware devices and configure a multisig (with or without a key on their machine)

Sjors committed:
rpc: add path argument to gethdkey

Sjors committed:
wallet: GetExtKey()

jonatack commented:
Thanks for working on this!

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -3258,3 +3258,34 @@ ScriptPubKeyMan* CWallet::AddWalletDescriptor(WalletDescriptor& desc, const Flat
+    auto desc_spk_man = dynamic_cast<DescriptorScriptPubKeyMan*>(spk_man);
+    CKey seed = desc_spk_man->GetKeys().begin()->second;
+    CExtKey master_key;
+    master_key.SetSeed(seed.begin(), seed.size());
```
In addition to being an ugly hack, this is also incorrect and does not actually produce the correct master key.

Rspigler commented:
Thanks so much for continuing to work on this! Love the use of BIP 87 :)

> Call `importdescriptors` which will allow the import if its own fingerprint is recognized (and after checking the xpub)

Yes, important to check the xpub also, not just the fingerprint, which is easy to spoof

> - export their own xpub and then import a descriptor...
> - import xpub(s) from other wallets, compose a multisig...
> - get xpub(s) from connected hardware devices...

I agree that the above 3 encompasses all use cases/needs for Bitcoin Core being a coordinator and a signer in a multisig setup.

> `importdescriptors` should treat an xpub as if it was an xpriv, after checking that it can derive the xpub from the seed (that matches the fingerprint); it needs to copy the seed from a sibling descriptor??

Can you explain this to me more?

As for the blank wallets, why do we need to be working from blank wallets, and if xprivs are derived from seeds, how can a blank wallet have a seed?

Sjors commented:
It depends on how we want to define `blank`. It would be empty except for a seed, so without any descriptors in it.

If a wallet has a seed and you call `importdescriptors` on it with some multisig descriptor, it should check the fingerprint(s) and path(s) mentioned in that descriptor. It should then derive the xpriv & xpub for that path, from the wallet seed, and check if it matches the descriptor `xpub`. In that case, she wallet should replace that xpub with the xpriv, as if the user provided it upon import. That way, if you want to sign a transaction, the wallet knows it can provide one of the signatures.

Rspigler commented:
Thank you, makes a lot of sense!

And what about sibling descriptors?

lsilva01 commented on test/functional/wallet_getxpub.py:
```diff
@@ -0,0 +1,64 @@
+        self.skip_if_no_sqlite()
+
+    def run_test(self):
+        self.nodes[0].createwallet(wallet_name="w1")
+        w1 = self.nodes[0].get_wallet_rpc("w1")
```
With a wallet different from the default one, I receive the error when sending to w2 address.
`test_framework.authproxy.JSONRPCException: Insufficient funds (-6)`

With the change below, it worked.

```suggestion
        w1 = self.nodes[0].get_wallet_rpc(self.default_wallet_name)
```

lsilva01 commented on test/functional/wallet_getxpub.py:
```diff
@@ -0,0 +1,64 @@
+        #       with getnewaddress.
+        w2_address = self.nodes[1].deriveaddresses(descsum_create(desc_receive), 0)[0]
+        w1.sendtoaddress(w2_address, 1)
+        self.nodes[0].generate(1)
```
I think `self.sync_all()` is necessary for w2 to know that it received money.

```suggestion
        self.nodes[0].generate(1)
        self.sync_all()
```

lsilva01 commented on test/functional/wallet_getxpub.py:
```diff
@@ -0,0 +1,64 @@
+
+        self.log.info("Spend from xpub")
+        w1_address = self.nodes[0].getnewaddress()
+        w2.sendtoaddress(w1_address, 0.1)
```
Just confirming that I got the error: `test_framework.authproxy.JSONRPCException: Signing transaction failed (-6)`

lsilva01 reviewed (commented):
Code Review ACK https://github.com/bitcoin/bitcoin/pull/22341/commits/1ff7a25d5af28481537f146632f57a3f20f53420

mjdietzx commented:
Concept ACK

I'll be sure to update the tests/docs in #22067 to use this when this PR gets merged

fjahr commented:
Concept ACK

Zero-1729 commented:
Concept ACK

achow101 committed:
descriptor: Be able to get the pubkeys involved in a descriptor

Sjors committed:
rpc: ParsePathBIP32 helper

Sjors commented:
Ok, this is now ready for review.

> I'll be sure to update the tests/docs in #22067 to use this when this PR gets merged

I might also take a stab at this, once this PR (and its ancestor) is rebased to include it.

Sjors committed:
test: use getxpub in wallet_multisig_descriptor_psbt.py

Sjors committed:
doc: use gethdkey in basic multisig descriptor example

Sjors commented:
Rebased and updated the multisig descriptor test and example documentation to use `getxpub`. Unfortunately we still have to reuse existing wallet keys, e.g. `m/44h/1h/0h`, because `importdescriptors` can't derive private keys yet.

JeremyRubin commented:
concept ACK this + using core as a signer like this.


One also cool idea would be is if, during PSBT signing code, we temporarily treat all descriptors there as "ours"?

Sjors commented:
@JeremyRubin afaik descriptors are always treated as "ours", both in private key wallets and watch-only wallets. The distinction between watch-only and regular addresses was dropped with the introduction descriptor wallets.

mjdietzx commented:
@Sjors in the description of this PR you say "The current draft implementation is not usable." what specifically are you referring to by that?

Sjors commented:
@mjdietzx I think that refers to an older version of this PR that used a hideous hack, but the current code should be fine to review. I deleted that remark for the description.

mjdietzx commented:
Code review ACK 8042053d0ddd662f654207d6554fb8f1e806f729

mjdietzx reviewed (commented):
Tested ACK 8042053d0ddd662f654207d6554fb8f1e806f729

I went through a variety of edge-cases checking that `getxpub` works as expected for error cases in various test wallets I have setup

Sjors commented on doc/descriptors.md:
```diff
@@ -155,8 +155,8 @@ The basic steps are:
      the participant's signer wallet. Avoid reusing this wallet for any purpose other than signing transactions from the
-     corresponding multisig we are about to create. Hint: extract the wallet's xpubs using `listdescriptors` and pick the one from the
-     `pkh` descriptor since it's least likely to be accidentally reused (legacy addresses)
+     corresponding multisig we are about to create. Hint: extract the wallet's xpubs using `getxpub m/44h/0h/0h` (`m/44h/1h/0h` for testnet and signet). This reuses keys from the
```
If using `h` doesn't work that's a bug. I'll look into that (probably after the next rebase).

Sjors replied:
The test cover both the use of `'` and `h`. You're not allowed to mix them in the same descriptor though.

Rspigler commented:
Tested commit 0ecd8a338e7f015e711cdc05ba005cdd322545e1

Got this error when running `make`:

>   CXX      wallet/libbitcoin_wallet_a-keyman.o
> In file included from ./script/signingprovider.h:11,
>                  from ./wallet/crypter.h:10,
>                  from wallet/keyman.cpp:9:
> ./script/keyorigin.h: In member function ‘std::optional<std::pair<CExtPubKey, KeyOriginInfo> > wallet::KeyManager::GetExtPubKey(std::vector<unsigned int>) const’:
> ./script/keyorigin.h:11:8: warning: ‘origin’ may be used uninitialized in this function [-Wmaybe-uninitialized]
>    11 | struct KeyOriginInfo
>       |        ^~~~~~~~~~~~~
> wallet/keyman.cpp:90:19: note: ‘origin’ was declared here
>    90 |     KeyOriginInfo origin;
>       |                   ^~~~~~
>   CXX      wallet/libbitcoin_wallet_a-load.o
> 

Still was able to compile

Tests all pass

Create a wallet with private keys disabled

1. Create a blank wallet with a seed (single key descriptor that is not active) as described in OP.

`importdescriptors "[{ \"desc\": \"pk(0279be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798)#gn28ywm7\", \"active\": false, \"timestamp\": \"now\" }]"`

> [
>   {
>     "success": true
>   }
> ]

2. Call `getxpub m/87h/0h/0h`

`getxpub m/87h/0h/0h`

>This wallet does not have an active master extended key (code -4)

Hm, ok - I'll repeat with a descriptor wallet, but won't be able to test the `getnewaddress` flow

1. Create a regular descriptor wallet
2. Call `getxpub m/87h/0h/0h`

`getxpub m/87h/0h/0h`

> {
>   "xpub": "xpub6BxZHpQ2UJZML59GZRPirFsWFJDqcYn4jcM6zm5qHUZnvuCC7xmb86meQywCbKuGEiuh3XvrvG1bhRiQDsYGmGRhFkMRNqYnBVtJYPCVt43",
>   "fingerprint": "5d340155",
>   "origin": "[5d340155/87'/0'/0']"
> }

Test importdescriptors:

`importdescriptors "[{ \"desc\": \"wpkh([5d340155/87'/0'/0]xpub6BxZHpQ2UJZML59GZRPirFsWFJDqcYn4jcM6zm5qHUZnvuCC7xmb86meQywCbKuGEiuh3XvrvG1bhRiQDsYGmGRhFkMRNqYnBVtJYPCVt43/0/*)#zc6l5c3g\", \"timestamp\": \"now\" }]"`

> [
>   {
>     "success": false,
>     "error": {
>       "code": -4,
>       "message": "Cannot import descriptor without private keys to a wallet with private keys enabled"
>     },
>     "warnings": [
>       "Range not given, using default keypool range"
>     ]
>   }
> ]

Not sure how to proceed from here

Sjors commented:
> Got this error when running make:

@Rspigler are you sure that warning is not related to #23417?

I clarified the description a bit: the steps there don't currently work.

Rspigler commented:
I am able to compile and run #23417 without issue.

But still get this error when compiling this PR:

> CXX      wallet/libbitcoin_wallet_a-keyman.o
> In file included from ./script/signingprovider.h:11,
>                  from ./wallet/crypter.h:10,
>                  from wallet/keyman.cpp:9:
> ./script/keyorigin.h: In member function ‘std::optional<std::pair<CExtPubKey, KeyOriginInfo> > wallet::KeyManager::GetExtPubKey(std::vector<unsigned int>) const’:
> ./script/keyorigin.h:11:8: warning: ‘origin’ may be used uninitialized in this function [-Wmaybe-uninitialized]
>    11 | struct KeyOriginInfo
>       |        ^~~~~~~~~~~~~
> wallet/keyman.cpp:90:19: note: ‘origin’ was declared here
>    90 |     KeyOriginInfo origin;
>       |                   ^~~~~~
>   CXX      wallet/libbitcoin_wallet_a-load.o
> 

I understand the PR better now, thank you for clarifying the description.

create a regular descriptor wallet
`listdescriptors`:
...(listing taproot descriptors only):

> 
> {
>       "desc": "tr([3e89a5ab/86'/0'/0']xpub6DQdRUJdQPzZmkMkSANpS1tEfhM48yMoSXnacafKyCGdBCb8Ky6TUzygGsGF63bwnncozQsVYBAebo3BEqCLuZiPnh7EpCGEWwL13ap99jk/0/*)#p5us77gk",
>       "timestamp": 1657730232,
>       "active": true,
>       "internal": false,
>       "range": [
>         0,
>         999
>       ],
>       "next": 0
>     },
>     
>     {
>       "desc": "tr([3e89a5ab/86'/0'/0']xpub6DQdRUJdQPzZmkMkSANpS1tEfhM48yMoSXnacafKyCGdBCb8Ky6TUzygGsGF63bwnncozQsVYBAebo3BEqCLuZiPnh7EpCGEWwL13ap99jk/1/*)#sqe3rtcw",
>       "timestamp": 1657730234,
>       "active": true,
>       "internal": true,
>       "range": [
>         0,
>         999
>       ],
>       "next": 0
>     }


`getxpub "m/86'/0'/0'"`

> {
>   "xpub": "xpub6DQdRUJdQPzZmkMkSANpS1tEfhM48yMoSXnacafKyCGdBCb8Ky6TUzygGsGF63bwnncozQsVYBAebo3BEqCLuZiPnh7EpCGEWwL13ap99jk",
>   "fingerprint": "3e89a5ab",
>   "origin": "[3e89a5ab/86'/0'/0']"
> }


'listdescriptors'
(listing taproot only)

{
      "desc": "tr([3e89a5ab/86'/0'/0']xpub6DQdRUJdQPzZmkMkSANpS1tEfhM48yMoSXnacafKyCGdBCb8Ky6TUzygGsGF63bwnncozQsVYBAebo3BEqCLuZiPnh7EpCGEWwL13ap99jk/0/*)#p5us77gk",
      "timestamp": 1657730232,
      "active": true,
      "internal": false,
      "range": [
        0,
        999
      ],
      "next": 0
    },
    
    
    

  {
      "desc": "tr([3e89a5ab/86'/0'/0']xpub6DQdRUJdQPzZmkMkSANpS1tEfhM48yMoSXnacafKyCGdBCb8Ky6TUzygGsGF63bwnncozQsVYBAebo3BEqCLuZiPnh7EpCGEWwL13ap99jk/1/*)#sqe3rtcw",
      "timestamp": 1657730234,
      "active": true,
      "internal": true,
      "range": [
        0,
        999
      ],
      "next": 0
    }


The taproot descriptors are the same 

:heavy_check_mark: 

tACK commit 0ecd8a338e7f015e711cdc05ba005cdd322545e1

Sjors commented:
> But still get this error when compiling this PR:

Ok, I'll look into that, probably after the next rebase.

Sjors commented:
Rebased. @Rspigler the uninitialized variable warning should be gone now.

Rspigler commented:
tACK commit f6133f093e3e305b305506b68499e1ac6cde726b

Yes, it is all good now

Sjors commented:
(trivial rebase, but not worth reviewing yet because there will be another one soon)

Sjors commented:
Rebased and fixed `nodiscard` warning introduced in #25642.

w0xlt reviewed (commented):
Approach ACK

aureleoules commented on src/rpc/util.cpp:
```diff
@@ -1092,6 +1093,15 @@ std::vector<CScript> EvalDescriptorStringOrObject(const UniValue& scanobject, Fl
     return ret;
 }
 
+std::vector<uint32_t> ParsePathBIP32(const std::string path)
```
```suggestion
std::vector<uint32_t> ParsePathBIP32(const std::string& path)
```

aureleoules commented on src/wallet/keyman.cpp:
```diff
@@ -0,0 +1,296 @@
+    return m_active_xpub;
+}
+
+std::optional<std::pair<CExtPubKey, KeyOriginInfo>> KeyManager::GetExtPubKey(const std::vector<uint32_t> path) const {
```
```suggestion
std::optional<std::pair<CExtPubKey, KeyOriginInfo>> KeyManager::GetExtPubKey(const std::vector<uint32_t>& path) const {
```

aureleoules reviewed (commented):
Concept ACK & LGTM

achow101 committed:
walletdb: Read and write activehdkey record

achow101 committed:
wallet: Load activehdkey and its private keys

The activehdkey record will now be loaded into the wallet as its own
member variable. Since the private keys will always be in existing
descriptors in the wallet, the key for the active hd key is also pulled
out of those descriptors on loading.

achow101 committed:
wallet: Store master key used for automatically generated descriptors

achow101 committed:
wallet: Retrieve active hd pubkey

achow101 committed:
wallet, rpc: Add gethdkey RPC

achow101 committed:
wallet: Implement GetActiveHDPrivkey

achow101 committed:
key: Add constructor for CExtKey that takes CExtPubKey and CKey

We often need to construct a CExtKey given an CExtPubKey and CKey, so
implement a constructor that does that for us.

achow101 committed:
wallet: Automatically upgrade a wallet to have global hd key

achow101 committed:
test: Test automatic upgrade of descriptor wallets to have hd key

Sjors commented:
Rebased on top of #26728 instead of #23417.

jonatack commented:
Re-concept ACK (not sure if it's ready for review).

Sjors commented:
@jonatack it was a somewhat sloppy rebase. I also need to review the new base PR. Will mark as draft.

achow101 committed:
tests: Test for gethdkey

achow101 committed:
wallet: Always set WALLET_FLAG_GLOBAL_HD_KEY for new wallets

achow101 committed:
wallet: Set global hd key for migrated wallets

Sjors commented:
Can't reproduce the CI failure, but will try again once the upstream PR is merged or rebased past #26076.

Sjors committed:
ParseHDKeypath: support h as hardened marker

S3RK commented:
The description now says the PR depends on itself :-)

achow101 committed:
wallet, rpc: Check WALLET_FLAG_GLOBAL_HD_KEY in gethdkey

achow101 committed:
test: Also do backwards compat test with encrypted wallets

Best reviewed with `git show -w`

achow101 committed:
test: Check that no unencrypted records persist after encrypting

furszy committed:
test: add coverage for re-opening a downgraded encrypted wallet on master

The test creates a wallet on master, downgrades and encrypts the wallet.
Then, it tries to open it again on master.

achow101 committed:
walletdb: Check that unencrypted and crypted keys are mutually exclusive

Sjors commented:
I plan to re-work this on top of  #29130 with a similar interface to `createwalletdescriptor`. That is, if you have a regular wallet with normal descriptors, it will Just Work(tm). Otherwise you need to specify which master key to use.

Initially I'll make a separate PR. If #26728 is closed in favor of the new approach in #29130 then this PR can be closed too.

ryanofsky commented:
> Initially I'll make a separate PR. If #26728 is closed in favor of the new approach in #29130 then this PR can be closed too.

It seems like you could just rebase this on top of #29130, since the description of the PR and commits aren't going to change that much, they will just be based on #29130 instead of #26728. Either way seems fine, though.

Sjors commented:
@ryanofsky building on #29130 introduces a new argument to specify which master key you intend to use.

Sjors commented:
Given that #26728 was closed, I'll just modify this PR to build on top of #29130.

(bit busy with stratum v2 stuff, but it will happen)

Sjors commented:
This seems more involved than a simple rebase. #29130 changes `gethdkey` to `gethdkeys` making it a less obvious choice to add a `path` argument to. We probably need a new RPC that's more similar to `createwalletdescriptor` in how it's called.

I probably won't have time for this in the near future, so I'm closing this as up for grabs.
